llvm-passes/ObfuscateVar/ObfuscateVar.cpp

```diff
@@ -0,0 +1,365 @@
+#include "llvm/IR/ValueMap.h"
+
+#include "llvm/Support/raw_ostream.h"
+
+#include "llvm/Transforms/Utils/BasicBlockUtils.h"
+#include "llvm/Transforms/IPO/PassManagerBuilder.h"
+
+#ifndef NDEBUG
+#include "llvm/IR/Verifier.h"
+#include "llvm/Support/Debug.h"
+#endif
+
+using namespace llvm;
+
+namespace {
+  typedef uint32_t typeInter;
+  typedef std::pair<Value*,Value*> typeMapValue;
+  typedef ValueMap<Value*,typeMapValue> typeMap;
+  
+  class ObfuscateVar : public BasicBlockPass {
+  public:
+  
+    static char ID;
+    ObfuscateVar() : BasicBlockPass(ID) {}
+
+    virtual bool runOnBasicBlock(BasicBlock &BB) override {
+      /* 
+         From https://www.cs.ox.ac.uk/files/2936/RR-10-02.pdf:
+         Given a variable X, we split it such as
+         - A = X%10
+         - B = X//10
+       
+         Z=X+Y will be transform into :
+         - Z_A = (X_A+Y_A) mod 10
+         - Z_B = { 10*(X_B+Y_B)+(X_A+Y_A) } div 10
+
+         To rebuild the variable : 
+         X = 10*X_B+X_A
+
+      */
+    
+      for (typename BasicBlock::iterator I = BB.getFirstInsertionPt(),end = BB.end(); I != end; ++I) {
+      
+        Instruction &Inst = *I;
+        bool isVolatile = false;
+      
+        if(isValidInstForSplit(Inst)) {//Check if the instruction can be splited
+        
+          for (size_t i=0; i < Inst.getNumOperands(); ++i) {
+            if (Value *V = isValidCandidateOperand(Inst.getOperand(i))) {
+
+              if(!isSplited(V)){//Check if the operand is splited
+                dbgs() << *V << " isn't splitted\n";
+                splitVariable(V,Inst);
+                //dbgs() << *V << " is now splited\n";
+                
+              }
+
+            }//isValidCandidateOperand
+          }//for
+
+          if(BinaryOperator *Binop = dyn_cast<BinaryOperator>(&Inst)) {
+
+            //Get the operator's operands
+            Value *op0 = parseOperand(Binop->getOperand(0));
+            Value *op1 = parseOperand(Binop->getOperand(1));
+           
+            if(!(isSplited(op0) && isSplited(op1))){//Check if the two operands are splited. It should be ok due to the previous loop
+              dbgs() << "Error : operands aren't splited !\n";
+              break;
+            }
+           
+            switch(Binop->getOpcode()){
+              
+            case Instruction::Add :{//Split add instruction
+
+              dbgs() << "Split ADD instruction\n";
+              
+              IRBuilder<> Builder(&Inst);
+              Type *IntermediaryType = IntegerType::get(Inst.getParent()->getContext(),sizeof(typeInter)*8);//32bits
+    
+              Constant *C10 = ConstantInt::get(IntermediaryType,10,false);
+              
+              // Get X_A and X_B from the first operand
+              typeMap::iterator it=varsRegister.find(op0); 
+              
+              Value *op0_A = Builder.CreateLoad(it->second.first,isVolatile);
+              Value *op0_B = Builder.CreateLoad(it->second.second,isVolatile);
```

Comment:
You're still confusing load/stores and regular values. A load is meant to load a value from the memory, there's no reason you would be keeping pointers instead of actual values in the varsRegister map.